### PR TITLE
Initial Rule Editor

### PR DESCRIPTION
**Pull request by aounhaider1**
New user interface for working with rules. Allows to add, update and delete rules. Also allows to configure multiple triggers/actions/conditions in a rule.

### CODE REVIEW

File: extensions/ui/org.eclipse.smarthome.ui.paper/web/partials/rules.html [hunk 1]
```diff
@@ -58,8 +58,195 @@
...
+                    </div>
+                    <md-button class="md-fab md-icon-add"
+                        ng-click="openNewConditionDialog($event)" aria-label="Add"></md-button>
+
+                </div>
+            </div>
+        </div>
+    </div>
+
+	<!-- ITEMIS AG -->
```
Comment (maggu2810) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web/partials/rules.html]: What the intention of this comment?

File: extensions/ui/org.eclipse.smarthome.ui.paper/web/partials/rules.html [hunk 2]
```diff
@@ -106,23 +293,7 @@
...
-			</div>
-		</div>
-		<div class="container">
-			<md-input-container> <label>Script</label> <textarea
-				class="mono" name="script" ng-model="script"></textarea> </md-input-container>
-		</div>
-	</div>
+	
+	
+	<!-- ITEMIS AG -->
```
Comment (maggu2810) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web/partials/rules.html]: What the intention of this comment?

**Comment by aounhaider1** [comment]
To just differentiate between the code blocks written by me and the person before. In the above case, the code block should have been removed.

**Comment by maggu2810** [comment]
IMHO such comments should be removed before it will be merged.
We could use "git blame" if we want to find the author of code lines.

Your last commit does not contain a Signed-off-by line, but it is no problem as the commits should be squashed (after review).

> Also allows to configure multiple triggers/actions/conditions in a rule. 

If I press on the + of the Tiggers / Actions / Conditions there is no reaction. Should this already work?

**Comment by aounhaider1** [comment]
Apologies for my ignorance about git.
Yes it should be working. Did you do a maven install on the 'paperui' project? Currently the bower components are not directly tracked so you either have to do a maven install 'paperui' or just a bower install from the command line inside the project where bower.json is located (you would need 'node'). I will confirm from Kai, if it should be left this way or should the bower components be tracked.

**Comment by kaikreuzer** [comment]
> To just differentiate between the code blocks written by me and the person before

Then it should have been "Deutsche Telekom AG" :-)
But anyhow, @maggu2810 is right: This is one project with shared code ownership, there is no "mine" and "their" code - so feel free to change whatever you think makes sense, and the maintainers will make sure that it is ok before merging it.

**Comment by kaikreuzer** [comment]
> I will confirm from Kai, if it should be left this way or should the bower components be tracked.

Yes, as these are a lot of files (and Javascript is like source code), it is set up as it is - the files are retrieved through bower and are not part of the repo. What might be worth though is to check if we could add a step in the IDE setup that triggers bower and downloads the files.

**Comment by maggu2810** [comment]
I changed to "smarthome/extensions/ui/org.eclipse.smarthome.ui.paper" and run "mvn clean install". But it is still not working.
So I removed the directories that content is fetched ("rm -rf node/ node_modules/ web/bower_components/") and run "mvn clean install" again.

But it seems there is still some missing on my setup...
I used the "SmartHome Runtime" launch target in the Eclipse IDE and added the Paper UI.
Are there some requirements (another bundle) that needs to be activated in the launch config to get the three buttons working?

**Comment by maggu2810** [comment]
```text
Error: event is not defined
capp</$scope.openDialog@http://127.0.0.1:8080/ui/js/controllers.rules.js:13:6
capp</$scope.openNewActionDialog@http://127.0.0.1:8080/ui/js/controllers.rules.js:88:7
kb.prototype.functionCall/<@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:200:301
Dc[c]</<.compile/</</f@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:217:74
Pe/this.$get</n.prototype.$eval@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:126:246
Pe/this.$get</n.prototype.$apply@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:126:471
Dc[c]</<.compile/</<@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:217:124
lf/c@http://127.0.0.1:8080/ui/bower_components/angular/angular.min.js:32:384
 angular.min.js:102:396
```

**Comment by maggu2810** [comment]
I can set "targetEvent: null" (don't know what is target event used for) and I get an dialog for "+ Actions" and "+ Conditions".

If I press "+ Triggers" I get "ERROR: 404 - ProxyServlet: /ui/partials/dialog.addTrigger.html".

**Comment by aounhaider1** [comment]
Ok. I am able to reproduce the same error in Firefox. Can you confirm if you are also using Firefox and if its working in other browsers?

**Comment by maggu2810** [comment]
Yes, I am using Firefox.

My current diff:

```diff
diff --git a/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.rules.js b/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.rules.js
index eec29ae..82dc67c 100644
--- a/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.rules.js
+++ b/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.rules.js
@@ -10,7 +10,6 @@ var capp=angular
                                $mdDialog.show({
                                        controller : ctrl,
                                        templateUrl : url,
-                                       targetEvent : event,
                                        hasBackdrop : true,
                                        locals:params
                                });
@@ -77,11 +76,11 @@ var capp=angular
                                        };
 
                                        $scope.openNewTriggerDialog = function(event) {
-                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addTrigger.html',{nshow:false,module:{label:"",type:"",id:""},ruleID:$scope.path[3]?$scope.path[3]:""});
+                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addtrigger.html',{nshow:false,module:{label:"",type:"",id:""},ruleID:$scope.path[3]?$scope.path[3]:""});
                                        }
                                        
                                        $scope.openUpdateTriggerDialog = function(event,trigger) {
-                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addTrigger.html',{nshow:true,module:trigger,ruleID:$scope.path[3]});
+                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addtrigger.html',{nshow:true,module:trigger,ruleID:$scope.path[3]});
                                        }
 
                                        $scope.openNewActionDialog = function(event) {
diff --git a/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/rule.update.js b/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/rule.update.js
index d1b6d72..38fbcf5 100644
--- a/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/rule.update.js
+++ b/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/rule.update.js
@@ -45,11 +45,11 @@ capp.controller(
                                                        };
 
                                                        $scope.openNewTriggerDialog = function(event) {
-                                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addTrigger.html',{nshow:false,module:{label:"",type:"",id:""},ruleID:$scope.path[3]?$scope.path[3]:""});
+                                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addtrigger.html',{nshow:false,module:{label:"",type:"",id:""},ruleID:$scope.path[3]?$scope.path[3]:""});
                                                        }
                                                        
                                                        $scope.openUpdateTriggerDialog = function(event,trigger) {
-                                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addTrigger.html',{nshow:true,module:trigger,ruleID:$scope.path[3]});
+                                                               $scope.openDialog('addTriggerDialogController','partials/dialog.addtrigger.html',{nshow:true,module:trigger,ruleID:$scope.path[3]});
                                                        }
 
                                                        $scope.openNewActionDialog = function(event) {
```

**Comment by maggu2810** [comment]
Using my changes it works with Chrome, too.
I don't it have to test it without that changes because IMHO it have to work with Firefox :wink: 

All fine? Do you need some further feedback?

**Comment by aounhaider1** [comment]
```
-                                       targetEvent : event,
```
Yes, its a mistake that I made while refactoring. The target event is not passed to the function. And I suppose it will work fine if we remove the initialisation but I will still look into it, if its the right way or not, on Monday.

**Comment by maggu2810** [comment]
Thank you

**Comment by kaikreuzer** [comment]
ftr: some refactoring and feature improvements are on the way - plan is to finalize this PR early next week to have an initial UI for rules (with of course a lot of room for further improvements).

File: extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.module.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
+capp.controller(
```
Comment (sja) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.module.js]: You rely on a global variable *capp* here. This is defined at controllers.rules.js, which create a loading order dependency. Better way: `angular.module('PaperUI.controllers.rules').controller('addModuleDialogController', ...)` to get the module after all JS is loaded.

**Comment by sja** [comment]
@kaikreuzer I did a PR to the PR, maybe that will help. Basically the code looks good.

**Pull request merged**